server/datastore/mysql/hosts.go

```diff
@@ -712,9 +712,19 @@ func (ds *Datastore) EnrollHost(ctx context.Context, osqueryHostID, nodeKey stri
 // LoadHostByNodeKey loads the whole host identified by the node key.
 // If the node key is invalid it returns a NotFoundError.
 func (ds *Datastore) LoadHostByNodeKey(ctx context.Context, nodeKey string) (*fleet.Host, error) {
-	sqlStatement := `SELECT * FROM hosts WHERE node_key = ?`
+	// LoadHostByNodeKey is one of the most used queries (every time a host checks in),
+	// thus prepare such query once.
+	ds.prepLoadHostQueryOnce.Do(func() {
+		var err error
+		//nolint ds.loadHostStmt is closed in Datastore.Close
+		ds.loadHostStmt, err = sqlx.PreparexContext(ctx, ds.reader, `SELECT * FROM hosts WHERE node_key = ?`)
+		if err != nil {
+			panic(fmt.Sprintf("prepare LoadHostByNodeKey statement: %s", err))
+		}
+	})
+
 	var host fleet.Host
-	switch err := sqlx.GetContext(ctx, ds.reader, &host, sqlStatement, nodeKey); {
+	switch err := ds.loadHostStmt.GetContext(ctx, &host, nodeKey); {
```

Comment:
Note that a prepared statement must still be re-prepared for each connection (https://pkg.go.dev/database/sql#Stmt). This is fine and still likely an improvement, just mentioning it so we're not surprised if looking at specific metrics we still see it being prepared multiple times.

Reply:
Yes, that's fine. We should be reusing mysql connections across endpoint connections, so it detaches this time from the amount of endpoints calling in (in terms of complexity class).

Reply:
Ah, TIL, thanks.
```
[...]
If a Stmt is prepared on a DB, it will remain usable for the lifetime of the DB. When the Stmt needs to execute on a new underlying connection, it will prepare itself on the new connection automatically.
```